### DISCUSSION TITLE
Refactor unit tests so that each is contained in a dedicated function.

### DISCUSSION THREAD

**Discussion by Tasqu** [spine-tools/SpineOpt.jl#339] [Ideas]
SpineOpt already has a number of tests, and running them all takes a while. This can be a pain when trying to debug single unit tests after some changes break them.

According to @manuelma, each unit test should be self-contained, so that they don't rely on each other. If that's indeed the case, it should be possible to separate each unit test into a dedicated function, making it possible to call only that test when debugging.

Full testing could then be implemented using a master function that calls all the individual tests in sequence.

**Comment by Tasqu** [comment]
Hmm, this doesn't seem practical on the level I was hoping. A lot of the tests rely on common data defined at the beginning of each `@testset`, so separating e.g. each `constraint` unit test into something callable on its own would require redefining that common data over and over again, which is not ideal.

Each larger `@testset` could still be separated into its dedicated function for slightly more modularity. Although, I'm not sure how these functions should be accessed, since the tests are currently not a part of `SpineOpt.jl` directly. I suppose the tests would have to be converted into a sub-module in this case?

**Reply by Tasqu** [reply]
In conclusion, I'm not sure if it would be worth it, at least for now.

**Comment by clizbe** [comment]
@Tasqu Can we close this discussion?

**Comment by Tasqu** [comment]
@clizbe I'd say so. Would be nice, but definitely not a priority and potentially not even worth the effort.
